contracts/0.8.25/vaults/VaultFactory.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: GPL-3.0
+
+import {UpgradeableBeacon} from "@openzeppelin/contracts-v5.0.2/proxy/beacon/UpgradeableBeacon.sol";
+import {BeaconProxy} from "@openzeppelin/contracts-v5.0.2/proxy/beacon/BeaconProxy.sol";
+import {Clones} from "@openzeppelin/contracts-v5.0.2/proxy/Clones.sol";
+
+import {IStakingVault} from "./interfaces/IStakingVault.sol";
+
+pragma solidity 0.8.25;
+
+interface IDelegation {
+    struct InitializationParams {
+        uint256 managementFee;
+        uint256 performanceFee;
+        address manager;
+        address operator;
+    }
+
+    function DEFAULT_ADMIN_ROLE() external view returns (bytes32);
+
+    function MANAGER_ROLE() external view returns (bytes32);
+
+    function OPERATOR_ROLE() external view returns (bytes32);
+
+    function LIDO_DAO_ROLE() external view returns (bytes32);
+
+    function initialize(address admin, address stakingVault) external;
+
+    function setManagementFee(uint256 _newManagementFee) external;
+
+    function setPerformanceFee(uint256 _newPerformanceFee) external;
+
+    function grantRole(bytes32 role, address account) external;
+
+    function revokeRole(bytes32 role, address account) external;
+}
+
+contract VaultFactory is UpgradeableBeacon {
+    address public immutable delegationImpl;
+
+    /// @param _owner The address of the VaultFactory owner
+    /// @param _stakingVaultImpl The address of the StakingVault implementation
+    /// @param _delegationImpl The address of the Delegation implementation
+    constructor(
+        address _owner,
+        address _stakingVaultImpl,
+        address _delegationImpl
+    ) UpgradeableBeacon(_stakingVaultImpl, _owner) {
+        if (_delegationImpl == address(0)) revert ZeroArgument("_delegation");
+
+        delegationImpl = _delegationImpl;
+    }
+
+    /// @notice Creates a new StakingVault and Delegation contracts
+    /// @param _stakingVaultParams The params of vault initialization
+    /// @param _initializationParams The params of vault initialization
+    function createVault(
+        bytes calldata _stakingVaultParams,
+        IDelegation.InitializationParams calldata _initializationParams,
+        address _lidoAgent
+    ) external returns (IStakingVault vault, IDelegation delegation) {
+        if (_initializationParams.manager == address(0)) revert ZeroArgument("manager");
+        if (_initializationParams.operator == address(0)) revert ZeroArgument("operator");
+
+        vault = IStakingVault(address(new BeaconProxy(address(this), "")));
+
+        delegation = IDelegation(Clones.clone(delegationImpl));
+
+        delegation.initialize(address(this), address(vault));
+
+        delegation.grantRole(delegation.LIDO_DAO_ROLE(), _lidoAgent);
+        delegation.grantRole(delegation.MANAGER_ROLE(), _initializationParams.manager);
+        delegation.grantRole(delegation.OPERATOR_ROLE(), _initializationParams.operator);
+        delegation.grantRole(delegation.DEFAULT_ADMIN_ROLE(), msg.sender);
```

Comment:
scary though, let's add a comment that those roles above have their admins

Reply:
:warning: let's not rely on `msg.sender` and use input parameters instead to make the deployments automated without pain (suggested `defaultAdmin` above to replace `msg.sender`)